### PR TITLE
[feature/APPC-1163] - Gamification details screen update

### PR DESCRIPTION
**Pull request by joaopmgouveia**
**What does this PR do?**

   This PR implements changes to the "How it works" gamification screen by turning it into a BottomSheet. Some modifications were also made to the main gamification screen.

"My level" screen: https://zpl.io/aX56OEj
"How it works" screen: https://zpl.io/2jMgNo4

**Database changed?**

   No

**Where should the reviewer start?**

- RewardsLevelActivity.kt
- MyLevelFragment.kt
- HowItWorksFragment.kt

**How should this be manually tested?**

  Flow on how to test this or QA Tickets related to this use-case: [APPC-1163](https://aptoide.atlassian.net/browse/APPC-1163)

**What are the relevant tickets?**

  Tickets related to this pull-request: [APPC-1163](https://aptoide.atlassian.net/browse/APPC-1163)

**Questions:**

There will be another PR which will deal with the possibility of implementing offline first when entering the gamification screen without internet connection

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

### CODE REVIEW

**Review by JDSM01 — CHANGES_REQUESTED** [review]
Remove unused How It works fragment and logic
Remove how it works info button visibility logic

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt [hunk 1]
```diff
@@ -29,15 +35,18 @@ class HowItWorksFragment : DaggerFragment(), HowItWorksView {
   private lateinit var presenter: HowItWorksPresenter
   private lateinit var gamificationView: GamificationView
 
+  private lateinit var bonusEarnedTextView: TextView
+  private lateinit var totalSpendTextView: TextView
+  private lateinit var nextLevelFooter: TextView
+  private lateinit var bonusEarnedSkeleton: LottieAnimationView
+  private lateinit var totalSpendSkeleton: LottieAnimationView
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt]: Not needed

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt [hunk 2]
```diff
@@ -48,25 +57,88 @@ class HowItWorksFragment : DaggerFragment(), HowItWorksView {
...
+
+  private fun removeDecimalPlaces(value: Double): String {
+    val splitValue = value.toString()
+        .split(".")
+    return splitValue[0]
+  }
+
+  private fun highlightCurrentLevel(levelTextView: TextView, messageTextView: TextView,
+                                    bonusTextView: TextView) {
+    val currentLevelColour = Color.parseColor("#001727")
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt]: This color has an entry on the colors.xml use it instead

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt [hunk 3]
```diff
@@ -48,25 +57,88 @@ class HowItWorksFragment : DaggerFragment(), HowItWorksView {
...
       view.findViewById<ImageView>(R.id.ic_level)
           .setImageResource(levelResourcesMapper.mapDarkIcons(level))
       (fragment_gamification_how_it_works_levels_layout as LinearLayout).addView(view)
+      if (level.level == currentLevel) {
+        highlightCurrentLevel(levelTextView, messageTextView, bonusTextView)
+      }
     }
   }
 
+  override fun showPeekInformation(totalSpend: BigDecimal, bonusEarned: FiatValue) {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt:78]: Interface has the wrong variable name

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt [hunk 4]
```diff
@@ -48,25 +57,88 @@ class HowItWorksFragment : DaggerFragment(), HowItWorksView {
...
+    totalSpendTextView.text = getString(R.string.gamification_how_table_a2, totalSpendRounded)
+
+    bonusEarnedSkeleton.visibility = View.INVISIBLE
+    totalSpendSkeleton.visibility = View.INVISIBLE
+    bonusEarnedTextView.visibility = View.VISIBLE
+    totalSpendTextView.visibility = View.VISIBLE
+  }
+
+  override fun showNextLevelFooter(userStatus: UserRewardsStatus) {
+    if (userStatus.level == 4) {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt]: The 4 could be a static variable. I know it’s not being used anywhere else, but for future purposes.

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksPresenter.kt [hunk 5]
```diff
@@ -32,24 +38,63 @@ class HowItWorksPresenter(private val view: HowItWorksView,
             .subscribeOn(networkScheduler)
             .observeOn(viewScheduler)
             .doOnSuccess {
-              view.showLevels(it)
+              view.showLevels(it.first, it.second)
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksPresenter.kt]: Not from this PR, but can you put this everything on the same line

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksPresenter.kt [hunk 6]
```diff
@@ -32,24 +38,63 @@ class HowItWorksPresenter(private val view: HowItWorksView,
...
+    disposables.add(
+        Single.zip(gamification.getLevels(), gamification.getUserStatus(),
+            gamification.getLastShownLevel(),
+            Function3 { levels: Levels, userStats: UserStats, lastShownLevel: Int ->
+              mapToUserStatus(levels, userStats, lastShownLevel)
+            })
+            .observeOn(viewScheduler)
+            .doOnSuccess {
+              view.showNextLevelFooter(it)
+            }
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksPresenter.kt]: same as above

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt [hunk 7]
```diff
@@ -48,25 +57,88 @@ class HowItWorksFragment : DaggerFragment(), HowItWorksView {
...
     for (level in levels) {
       view = layoutInflater.inflate(R.layout.fragment_gamification_how_it_works_level,
           fragment_gamification_how_it_works_levels_layout, false)
-      view.findViewById<TextView>(R.id.message).text =
-          getString(R.string.gamification_how_table_a2
-              , level.amount)
-      view.findViewById<TextView>(R.id.bonus).text =
-          getString(R.string.gamification_how_table_b2, level.bonus.toString())
+      val levelTextView = view.findViewById<TextView>(R.id.level)
+      val messageTextView = view.findViewById<TextView>(R.id.message)
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/HowItWorksFragment.kt]: spendTextView maybe?

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelFragment.kt [hunk 8]
```diff
@@ -54,11 +56,17 @@ class MyLevelFragment : DaggerFragment(), MyLevelView {
...
+        .replace(R.id.gamification_fragment_container, HowItWorksFragment())
+        .commit()
+    return inflater.inflate(R.layout.fragment_gamification_my_level, container, false)
   }
 
   override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
     super.onViewCreated(view, savedInstanceState)
+    howItWorksBottomSheet =
+        BottomSheetBehavior.from(view.findViewById<View>(R.id.gamification_fragment_container))
+    backgroundFadeView = view.findViewById(R.id.background_fade_animation)
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelFragment.kt]: view.findViewById not needed

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelFragment.kt [hunk 9]
```diff
@@ -253,8 +243,20 @@ class MyLevelFragment : DaggerFragment(), MyLevelView {
...
+
+      override fun onSlide(bottomSheet: View, slideOffset: Float) {
+        backgroundFadeView.progress = slideOffset
+      }
+    })
+  }
+
   private fun animateLevelToLock(levelIcon: View?, levelText: TextView?) {
-    var icon: ImageView = levelIcon?.findViewById(R.id.level_active_icon) as ImageView
+    val icon: ImageView = levelIcon?.findViewById(R.id.level_active_icon) as ImageView
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelFragment.kt]: Clean not needed casts and null possibilities

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelFragment.kt [hunk 10]
```diff
@@ -303,16 +305,21 @@ class MyLevelFragment : DaggerFragment(), MyLevelView {
     level_description.text = getString(levelResourcesMapper.mapSubtitle(level))
     level_description.visibility = View.VISIBLE
     current_level.text =
-        getString(R.string.gamification_level_on_graphic, Integer.toString(level + 1))
+        getString(R.string.gamification_level_on_graphic, (level + 1).toString())
   }
 
-  private fun setLevelBonus(level: Int, text: String) {
+  private fun setLevelBonus(level: Int, text: String, isCurrentLevel: Boolean) {
+    val bonusLabel: Int = if (isCurrentLevel) {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelFragment.kt]: Int not needed

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelPresenter.kt [hunk 11]
```diff
@@ -12,21 +12,19 @@ import java.math.BigDecimal
 import java.math.RoundingMode
 
 class MyLevelPresenter(private val view: MyLevelView,
+                       private val gamificationView: GamificationView?,
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelPresenter.kt]: Change to gamificationActivity (variable name)

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelPresenter.kt [hunk 12]
```diff
@@ -50,6 +48,13 @@ class MyLevelPresenter(private val view: MyLevelView,
             .subscribe())
   }
 
+  private fun handleInfoButtonClick() {
+    disposables.add(
+        gamificationView!!.getInfoButtonClick()!!.doOnNext {
+          view.changeBottomSheetState()
+        }!!.subscribe())
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelPresenter.kt]: Change to gamificationView.let{disposables.add(it.getInfo...}

File: app/src/main/res/layout/fragment_gamification_how_it_works.xml [hunk 13]
```diff
@@ -2,141 +2,273 @@
...
+      app:layout_constraintTop_toBottomOf="@id/fragment_gamification_how_it_works_message"
+      />
+
+  <LinearLayout
+      android:id="@+id/fragment_gamification_how_it_works_levels_layout"
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_marginTop="8dp"
+      android:orientation="vertical"
+      android:visibility="visible"
```
Comment (JDSM01) [inline app/src/main/res/layout/fragment_gamification_how_it_works.xml]: Not needed

**Review by JDSM01 — CHANGES_REQUESTED** [review]
BottomSheet amounts are a little below of the required by the design

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelPresenter.kt [hunk 14]
```diff
@@ -48,10 +48,11 @@ class MyLevelPresenter(private val view: MyLevelView,
   }
 
   private fun handleInfoButtonClick() {
-    disposables.add(
-        gamificationView!!.getInfoButtonClick()!!.doOnNext {
-          view.changeBottomSheetState()
-        }!!.subscribe())
+    activity.let {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelPresenter.kt]: My bad, you need to use ? before the let and then remove the !!

File: app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelPresenter.kt [hunk 15]
```diff
@@ -48,10 +48,11 @@ class MyLevelPresenter(private val view: MyLevelView,
...
+      disposables.add(it!!.getInfoButtonClick()!!.doOnNext {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/gamification/MyLevelPresenter.kt]: And then use the !! on the return method of the fragment

**Review by JDSM01 — CHANGES_REQUESTED** [review]

**Review by JDSM01 — CHANGES_REQUESTED** [review]
- [ ] Sometimes when you enter the gamification screen the level graphic doesn’t show up. This happens more, but not always when in the transactions view, the transactions are still loading.

- [X] The loading from the how it works fragment is not using the theme colour

- [X] If you have network, go to the view (with a wallet level different from 1), go back, disable network, then go to the view, it shows level 1, but all the icons are highlighted and no lines between them

**Review by carvalho2707 — COMMENTED** [review]
Shouldn't we have the same style for all progress bars?

**Comment by carvalho2707** [comment]
> Shouldn't we have the same style for all progress bars?

This will be implemented in a different ticket

**Review by carvalho2707 — APPROVED** [review]

**Pull request merged**